consensus/state_processing/src/per_epoch_processing/single_pass.rs

```diff
@@ -369,10 +372,60 @@ pub fn process_epoch_single_pass<E: EthSpec>(
                 signature: deposit.signature,
             };
             if is_valid_deposit_signature(&deposit_data, spec).is_ok() {
-                state.add_validator_to_registry(
-                    deposit_data.pubkey,
-                    deposit_data.withdrawal_credentials,
-                    deposit_data.amount,
+                // If a validator with the same pubkey has already been added, do not add a new
+                // validator, just update the balance.
+                // Note: updating the `effective_balance` for the new validator is done after
+                // adding them to the state.
+                if let Some((_, validator_index)) = added_validators
+                    .iter()
+                    .find(|(pubkey, _)| *pubkey == deposit_data.pubkey)
+                {
+                    let balance = state.get_balance_mut(*validator_index)?;
+                    balance.safe_add_assign(deposit_data.amount)?;
+                } else {
+                    // Apply the new deposit to the state
+                    state.add_validator_to_registry(
+                        deposit_data.pubkey,
+                        deposit_data.withdrawal_credentials,
+                        deposit_data.amount,
+                        spec,
+                    )?;
+                    added_validators
+                        .push((deposit_data.pubkey, state.validators().len().safe_sub(1)?));
+                }
+            }
+        }
+        if conf.effective_balance_updates {
+            // Re-process effective balance updates for validators affected by top-up of new validators.
```

Comment:
I'm very sus about this. Seems to have fixed the deposits processing bug though
cc @michaelsproul

Reply:
For some added context, we want deposit processing to work when `process_pending_deposits` processes 2 deposits where deposit 1 is a new validator and deposit 2 is a topup for the new validator.
Since `process_effective_balance_updates` happens after `process_pending_deposits`, we need the effective balance update happening after adding the new validator

Reply:
I agree with this assessment of the situation. It's a bit gnarly that we need to do an effective balance update here outside the main single pass loop, but I think it's the best way.

I was thinking it might make sense to put this block of code after the logic for processing pending consolidations because that's the order used by the spec:

```python
process_pending_deposits(state)  # [New in Electra:EIP7251]
process_pending_consolidations(state)  # [New in Electra:EIP7251]
process_effective_balance_updates(state)  # [Modified in Electra:EIP7251]
```

However, the spec also requires that consolidations are for known validators (both source and target), so that should not interact at all with our newly added validators.

Reply:
The only other thing I would consider doing would be getting rid of the `added_validators` vec and just running effective balance updates for all the validators from `num_validators..` on wards.

Ideally we could even use something like `iter_from` that returns CoW pointers (`iter_cow_from`), however this doesn't exist in Milhouse yet.